src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -22,57 +23,66 @@ use super::{
 
 pub type ArcMutexDevice = Arc<Mutex<dyn Device>>;
 
+macro_rules! declare_index {
+    ($self:ident, $index:ident, $released_index:ident) => {{
+        let current_index = if let Some(index) = $self.$released_index.pop() {
+            index
+        } else {
+            $self.$index
+        };
+        $self.$index += 1;
+        Ok(current_index)
+    }};
+}
+
+macro_rules! release_index {
+    ($self:ident, $index:ident, $released_index:ident) => {{
+        $self.$released_index.push($index);
+        $self.$released_index.sort_by(|a, b| b.cmp(a));
+    }};
+}
+
 /// block_index and released_block_index are used to search an available block index
 /// in Sandbox.
+/// pmem_index and released_pmem_index are used to search an available pmem index
+/// in Sandbox.
 ///
-/// @block_driver to be used for block device;
-/// @block_index generally default is 1 for <vdb>;
+/// @pmem_index generally default is 0 for <pmem0>;
+/// @block_index generally default is 0 for <vda>;
+/// @released_pmem_index for pmem devices removed and indexes will released at the same time.
 /// @released_block_index for blk devices removed and indexes will released at the same time.
 #[derive(Clone, Debug, Default)]
 struct SharedInfo {
-    block_driver: String,
+    pmem_index: u64,
     block_index: u64,
+    released_pmem_index: Vec<u64>,
     released_block_index: Vec<u64>,
 }
 
 impl SharedInfo {
-    async fn new(hypervisor: Arc<dyn Hypervisor>) -> Self {
-        // get hypervisor block driver
-        let block_driver = match hypervisor
-            .hypervisor_config()
-            .await
-            .blockdev_info
-            .block_device_driver
-            .as_str()
-        {
-            // convert the block driver to kata type
-            VIRTIO_BLOCK_MMIO => KATA_MMIO_BLK_DEV_TYPE.to_string(),
-            VIRTIO_BLOCK_PCI => KATA_BLK_DEV_TYPE.to_string(),
-            _ => "".to_string(),
-        };
-
+    async fn new() -> Self {
         SharedInfo {
-            block_driver,
-            block_index: 1,
+            pmem_index: 0,
+            block_index: 0,
+            released_pmem_index: vec![],
             released_block_index: vec![],
         }
     }
 
-    // declare the available block index
-    fn declare_device_index(&mut self) -> Result<u64> {
-        let current_index = if let Some(index) = self.released_block_index.pop() {
-            index
-        } else {
-            self.block_index
-        };
-        self.block_index += 1;
+    fn declare_pmem_index(&mut self) -> Result<u64> {
+        declare_index!(self, pmem_index, released_pmem_index)
+    }
 
-        Ok(current_index)
+    fn declare_block_index(&mut self) -> Result<u64> {
+        declare_index!(self, block_index, released_block_index)
+    }
+
+    fn release_pmem_index(&mut self, index: u64) {
+        release_index!(self, index, released_pmem_index);
```

Comment:
IMO, the `declare_device_index` and `release_device_index` should be reserved and add an argument to  differentiate the NVDIMM type and BLOCK type, If so, dup codes will be reduced.

Reply:
thx, fixed